File: Yandex/Practicum/Algorightms_and_Data_Structures/4_sprint/L._МногоГоша/solution.py
import sys

MOD1: int = 100000003
MOD2: int = 100000019
MOD3: int = 100000037
q1: int = 100000009
q2: int = 100000007
q3: int = 100000021


def solution() -> None:
    """
    Time Complexity O(N)
    Memory Complexity O(N)
    """
    n, k = map(int, sys.stdin.readline().rstrip().split())
    L: str = sys.stdin.readline().rstrip()
    hashmap: dict[tuple[int, int, int], list[int, int]] = dict()
    first: bool = True

    cur_hash1: int = 0
    cur_hash2: int = 0
    cur_hash3: int = 0
    q_pow_k1: int = pow(q1, n, MOD1)
    q_pow_k2: int = pow(q2, n, MOD2)
    q_pow_k3: int = pow(q3, n, MOD3)

    for i in range(len(L)):
        cur_hash1 = (cur_hash1 * q1 + ord(L[i])) % MOD1
        cur_hash2 = (cur_hash2 * q2 + ord(L[i])) % MOD2
        cur_hash3 = (cur_hash3 * q3 + ord(L[i])) % MOD3

        if i >= n - 1:
            if i >= n:
                cur_hash1 = (cur_hash1 - ord(L[i - n]) * q_pow_k1) % MOD1
                cur_hash2 = (cur_hash2 - ord(L[i - n]) * q_pow_k2) % MOD2
                cur_hash3 = (cur_hash3 - ord(L[i - n]) * q_pow_k3) % MOD3
            triplet: tuple[int, int, int] = (cur_hash1, cur_hash2, cur_hash3)
            if triplet not in hashmap:
                hashmap[triplet] = [0, i - n + 1]
            hashmap[triplet][0] += 1
            if hashmap[triplet][0] == k:
                if not first:
                    sys.stdout.write(' ')
                sys.stdout.write(str(hashmap[triplet][1]))
                first = False


if __name__ == '__main__':
    solution()

File: Yandex/Practicum/Algorightms_and_Data_Structures/4_sprint/L._МногоГоша/solution.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <tuple>
#include <functional>

using namespace std;

const long long MOD1 = 100000003;
const long long MOD2 = 100000019;
const long long MOD3 = 100000037;
const long long q1 = 100000009;
const long long q2 = 100000007;
const long long q3 = 100000021;

struct TupleHash {
    template <class T>
    size_t operator()(const tuple<T, T, T>& t) const {
        auto hash1 = hash<T>()(get<0>(t));
        auto hash2 = hash<T>()(get<1>(t));
        auto hash3 = hash<T>()(get<2>(t));
        return hash1 ^ (hash2 << 1) ^ (hash3 << 2);
    }
};


void solution() {
    /*
    Time Complexity O(N)
    Memory Complexity O(N)
    */
    int n, k;
    std::cin >> n >> k;
    std::string L;
    std::cin >> L;
    bool seen = false;
    std::unordered_map<
        std::tuple<long long, long long, long long>,
        std::pair<int, int>,
        TupleHash
    > hashmap;
    long long cur_hash1 = 0, cur_hash2 = 0, cur_hash3 = 0;
    long long q_pow_k1 = 1, q_pow_k2 = 1, q_pow_k3 = 1;

    for (int i = 0; i < n; ++i) {
        q_pow_k1 = (q_pow_k1 * q1) % MOD1;
        q_pow_k2 = (q_pow_k2 * q2) % MOD2;
        q_pow_k3 = (q_pow_k3 * q3) % MOD3;
    }

    for (int i = 0; i < L.size(); ++i) {
        cur_hash1 = (cur_hash1 * q1 + (L[i] - 'a' + 1)) % MOD1;
        cur_hash2 = (cur_hash2 * q2 + (L[i] - 'a' + 1)) % MOD2;
        cur_hash3 = (cur_hash3 * q3 + (L[i] - 'a' + 1)) % MOD3;
        if (i >= n - 1) {
            if (i >= n) {
                cur_hash1 = (cur_hash1 - (L[i - n] - 'a' + 1) * q_pow_k1 % MOD1 + MOD1) % MOD1;
                cur_hash2 = (cur_hash2 - (L[i - n] - 'a' + 1) * q_pow_k2 % MOD2 + MOD2) % MOD2;
                cur_hash3 = (cur_hash3 - (L[i - n] - 'a' + 1) * q_pow_k3 % MOD3 + MOD3) % MOD3;
            }
            std::tuple<long long, long long, long long> triplet = {cur_hash1, cur_hash2, cur_hash3};
            if (!hashmap.count(triplet)) {
                hashmap[triplet] = {0, i - n + 1};
            }
            ++hashmap[triplet].first;
            if (hashmap[triplet].first == k) {
                if (seen) {
                    std::cout << " ";
                }
                std::cout << hashmap[triplet].second;
                seen = true;
            }
        }
    }
}

int main() {
    solution();
}